source/lib/rocprof-sys/library/amd_smi.cpp

```diff
@@ -417,25 +508,79 @@ data::post_process(uint32_t _dev_id)
             if(_settings.mem_usage)
                 TRACE_COUNTER("device_memory_usage",
                               counter_track::at(_dev_id, _idx.at(5)), _ts, _usage);
-            if(_settings.vcn_activity)
+            if(_settings.vcn_activity && !itr.m_xcp_metrics.empty())
             {
                 uint64_t idx = _idx.at(6);
-                for(const auto& temp : itr.m_vcn_metrics)
+                if(gpu::is_vcn_activity_supported(_dev_id))
                 {
-                    TRACE_COUNTER("device_vcn_activity", counter_track::at(_dev_id, idx),
-                                  _ts, temp);
-                    ++idx;
+                    // When VCN activity is supported, use the activity values
+                    for(size_t i = 0; i < itr.m_xcp_metrics[0].vcn_busy.size(); ++i)
+                    {
+                        TRACE_COUNTER("device_vcn_activity",
+                                      counter_track::at(_dev_id, idx), _ts,
+                                      itr.m_xcp_metrics[0].vcn_busy[i]);
+                        ++idx;
+                    }
+                }
+                else
+                {
+                    // When VCN activity is not supported, use XCP-specific busy values
+                    for(const auto& temp : itr.m_xcp_metrics)
+                    {
+                        for(size_t i = 0; i < temp.vcn_busy.size(); ++i)
+                        {
+                            TRACE_COUNTER("device_vcn_activity",
+                                          counter_track::at(_dev_id, idx), _ts,
+                                          temp.vcn_busy[i]);
+                            ++idx;
+                        }
+                    }
                 }
             }
-            if(_settings.jpeg_activity)
+
+            if(_settings.jpeg_activity && !itr.m_xcp_metrics.empty())
             {
                 uint64_t idx = _idx.at(7);
-                if(_settings.vcn_activity) idx += (itr.m_vcn_metrics.size() - 1);
-                for(const auto& temp : itr.m_jpeg_metrics)
+                // Calculate total VCN metrics to properly offset JPEG metrics index
+                if(_settings.vcn_activity)
                 {
-                    TRACE_COUNTER("device_jpeg_activity", counter_track::at(_dev_id, idx),
-                                  _ts, temp);
-                    ++idx;
+                    size_t total_vcn_metrics = 0;
+                    if(gpu::is_vcn_activity_supported(_dev_id))
+                    {
+                        total_vcn_metrics = itr.m_xcp_metrics[0].vcn_busy.size();
+                    }
+                    else
+                    {
+                        for(const auto& temp : itr.m_xcp_metrics)
+                            total_vcn_metrics += temp.vcn_busy.size();
+                    }
```

Comment:
```suggestion
                    for(const auto& metrics : itr.m_xcp_metrics)
                        total_vcn_metrics += metrics.vcn_busy.size();
```
We should be able to simplify this to the above. In gpus that don't support XCP, the size of m_xcp_metrics will just be 1.